### PR TITLE
Enhanced RDoc for String

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Treated:
- #slice!
- #sub
- #sub!
- #gsub
- #gsub!

### CODE REVIEW

File: string.c [hunk 1]
```diff
@@ -5477,13 +5482,16 @@ rb_pat_search(VALUE pat, VALUE str, long pos, int set_backref_str)
 
 /*
  *  call-seq:
- *     str.sub!(pattern, replacement)          -> str or nil
- *     str.sub!(pattern) {|match| block }      -> str or nil
+ *    sub!(pattern, replacement)   -> self or nil
+ *    sub!(pattern) {|match| ... } -> self or nil
+ *
+ *  Performs the specified substring replacement on +self+;
+ *  returns +self+ if any replacement occurred, +nil+ otherwise.
```
Comment (jeremyevans) [inline string.c:5489]: This should probably mention it only does a single substitution, similar to how `sub` is documented.

File: string.c [hunk 2]
```diff
@@ -11856,18 +11782,131 @@ rb_enc_interned_str_cstr(const char *ptr, rb_encoding *enc)
 }
 
 /*
- *  A String object holds and manipulates an arbitrary sequence of
- *  bytes, typically representing text or binary data. String objects may be
- *  created using String::new or as literals.
+ *  A \String object has an arbitrary sequence of bytes,
+ *  typically representing text or binary data.
+ *  A \String objects may be created using String::new or as literals.
```
Comment (jeremyevans) [inline string.c:11787]: This should either be `A \String object` or `\String objects`

File: string.c [hunk 3]
```diff
@@ -11856,18 +11782,131 @@ rb_enc_interned_str_cstr(const char *ptr, rb_encoding *enc)
...
  *
  *  String objects differ from Symbol objects in that Symbol objects are
- *  designed to be used as identifiers, instead of text or data.
+ *  commonly used as identifiers, instead of text or data.
```
Comment (jeremyevans) [inline string.c:11790]: Please change back from `commonly used` to `designed to be used`.  `commonly used` can imply that such usage is common but not a good idea.

File: string.c [hunk 4]
```diff
@@ -11856,18 +11782,131 @@ rb_enc_interned_str_cstr(const char *ptr, rb_encoding *enc)
...
+ *
+ *  Some \String methods modify +self+.
+ *  Typically, a method whose name ends with <tt>!</tt> modifies +self+
+ *  and returns +self+;
+ *  often a similarly named method (without the <tt>!</tt>)
+ *  returns a new string.
```
Comment (jeremyevans) [inline string.c:11796]: In general, if there exist both bang and non-bang version of methods, the bang! mutates and the non-bang! does not.  However, if a method without a bang can also mutate, such as `replace`.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks pretty good, only a few minor changes requested.

File: string.c [hunk 5]
```diff
@@ -5301,21 +5301,26 @@ rb_str_insert(VALUE str, VALUE idx, VALUE str2)
...
- *  deleted.
+ *  Removes the substring of +self+ specified by the arguments;
+ *  returns the removed substring.
  *
- *     string = "this is a string"
+ *  See String#[] for details about the arguments that specify the substring.
+ *
+ *  A few examples:
+
+ *     string = "This is a string"
```
Comment (nobu) [inline string.c:5317]: Changing the first string will make the result different.

Comment (BurdetteLamar) [reply string.c:5317]: > Changing the first string will make the result different.

Thanks, @nobu.  I'm not sure what you mean.  Is the example incorrect?

Comment (BurdetteLamar) [reply string.c:5317]: Now fixed, and thanks.

Comment (nobu) [reply string.c:5317]: Yes, seems correct.
I'm not sure why changed the example, though.

File: string.c [hunk 6]
```diff
@@ -5301,21 +5301,26 @@ rb_str_insert(VALUE str, VALUE idx, VALUE str2)
...
- *  Deletes the specified portion from <i>str</i>, and returns the portion
- *  deleted.
+ *  Removes the substring of +self+ specified by the arguments;
+ *  returns the removed substring.
  *
- *     string = "this is a string"
+ *  See String#[] for details about the arguments that specify the substring.
+ *
+ *  A few examples:
+
```
Comment (nobu) [inline string.c:5316]: Removed `*`?

File: string.c (hunk 5 above)
Comment (BurdetteLamar) [reply string.c:5317]: Just to make the sentence begin with a capital letter.  But on that principle might also have ended with a period.

**Review by jeremyevans — APPROVED** [review]
Looks good!

**Comment by mame** [comment]
Just FYI: I tweaked the document a bit https://github.com/ruby/ruby/commit/4b248e7994e68fa5075f62d4cbfb36514ffd7bfa

**Pull request merged**